### PR TITLE
refactor: migrate handle-data and local/src/index to typescript

### PR DESCRIPTION
**Pull request by LiKang6688**

### CODE REVIEW

File: src/handle-data.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
+import { stardust } from '@nebula.js/stardust';
+import { TableLayout, HyperCube, ExtendedNxDimensionInfo, ExtendedNxMeasureInfo, Column, PageInfo } from './types';
+
 const directionMap = {
   A: 'asc',
   D: 'desc',
+  N: 'asc',
```
Comment (haxxmaxx) [inline src/handle-data.ts:7]: why is this needed?

Comment (LiKang6688) [reply src/handle-data.ts:7]: good question. Becasue of the value of`info.qSortIndicator` in `directionMap[info.qSortIndicator]` is A or D or N. https://github.com/qlik-oss/sn-table/pull/586/files#diff-26fa10f48243e01a30f0b2eee16fa394336656c6e83d68066718bad1d1ccb1cdR55

File: src/handle-data.ts [hunk 2]
```diff
@@ -13,29 +17,31 @@ const MAX_CELLS = 10000;
...
   const { qDimensionInfo, qMeasureInfo } = layout.qHyperCube;
   const numDims = qDimensionInfo.length;
   const isDim = colIndex < numDims;
-  const info = isDim ? qDimensionInfo[colIndex] : qMeasureInfo[colIndex - numDims];
+  const info = (isDim ? qDimensionInfo[colIndex] : qMeasureInfo[colIndex - numDims]) as
+    | ExtendedNxMeasureInfo
+    | ExtendedNxDimensionInfo;
   const isHidden = info.qError?.qErrorCode === 7005;
-  const isLocked = info.qLocked;
+  const isLocked = isDim && qDimensionInfo[colIndex].qLocked;
```
Comment (haxxmaxx) [inline src/handle-data.ts:44]: why can't you use info here

File: src/handle-data.ts [hunk 3]
```diff
@@ -75,7 +81,16 @@ export function getTotalPosition(layout) {
...
-export default async function manageData(model, layout, pageInfo, setPageInfo) {
+export default async function manageData(
+  model: EngineAPI.IGenericObject,
+  layout: TableLayout,
+  pageInfo: PageInfo,
+  setPageInfo: stardust.SetStateFn<{
+    page: number;
+    rowsPerPage: number;
+    rowsPerPageOptions: number[];
+  }>
```
Comment (haxxmaxx) [inline src/handle-data.ts:92]: doesn't the function take a `PageInfo` object as argument?

File: src/handle-data.ts (hunk 2 above)
Comment (LiKang6688) [reply src/handle-data.ts:44]: as qLocked does not exist in qMeasureInfo

File: src/handle-data.ts [hunk 4]
```diff
@@ -99,13 +114,14 @@ export default async function manageData(model, layout, pageInfo, setPageInfo) {
...
+  const columns = columnOrder.map((colIndex: number) => getColumnInfo(layout, colIndex)).filter(Boolean) as Column[];
   const dataPages = await model.getHyperCubeData('/qHyperCubeDef', [
     { qTop: top, qLeft: 0, qHeight: height, qWidth: totalColumnCount },
   ]);
 
-  const rows = dataPages[0].qMatrix.map((r, rowIdx) => {
-    const row = { key: `row-${rowIdx}` };
+  const rows = dataPages[0].qMatrix.map((r, rowIdx: number) => {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const row = { key: `row-${rowIdx}` } as any;
```
[the remark this answers is not in the document]
Comment (LiKang6688) [reply src/handle-data.ts:124]: I cannot come up with one. any suggestions?

File: src/handle-data.ts (hunk 2 above)
Comment (haxxmaxx) [reply src/handle-data.ts:44]: but it will never be qMeasureInfo if isDim is true. Does
```suggestion
  const isLocked = isDim && info.qLocked;
```
work or does typescript complain?

Comment (LiKang6688) [reply src/handle-data.ts:44]: yes, typescript does not know that, so it complains

Comment (haxxmaxx) [reply src/handle-data.ts:44]:
```suggestion
  const isLocked = isDim && (info as ExtendedNxDimensionInfo).qLocked;
```

File: src/handle-data.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (haxxmaxx) [reply src/handle-data.ts:124]: I have a commit that I think would work

[the remark this answers is not in the document]
Comment (LiKang6688) [reply src/handle-data.ts:124]: which commit?

File: src/handle-data.ts (hunk 3 above)
Comment (LiKang6688) [reply src/handle-data.ts]: Fixed

File: src/handle-data.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (haxxmaxx) [reply src/handle-data.ts:120]: I have it locally. Something like this, types:
```
export interface TableCell {
  qText: string | undefined;
  qAttrExps: EngineAPI.INxAttributeExpressionValues;
  rowIdx: number;
  colIdx: number;
  isSelectable: boolean;
  rawRowIdx: number;
  rawColIdx: number;
  prevQElemNumber: number;
  nextQElemNumber: number;
}

export interface TableRow {
  // for the row key, string is needed
  [key: string]: TableCell | string;
}
```
and then in handle data:
```
    const row: TableRow = { key: `row-${rowIdx}` };
    columns.forEach((c, colIdx) => {
      const { qText, qAttrExps } = r[colIdx];
      row[c.id] = {
        qText,
        qAttrExps,
        rowIdx: rowIdx + top,
        ...
```

File: src/handle-data.ts [hunk 5]
```diff
@@ -13,29 +17,31 @@ const MAX_CELLS = 10000;
  * @param {Number} numDims
  * @returns dimensions and measures total cell values as strings
  */
-export function getTotalInfo(isDim, layout, colIndex, numDims) {
+export function getTotalInfo(isDim: boolean, layout: TableLayout, colIndex: number, numDims: number) {
```
Comment (haxxmaxx) [inline src/handle-data.ts:20]: have a look if the type for layout exists on stardust

File: src/types.ts [hunk 6]
```diff
@@ -1,4 +1,73 @@
 export interface Translator {
   get: (str: string, args?: string[]) => string;
+  add: (item: { id: string; locale: object }) => void;
```
Comment (haxxmaxx) [inline src/types.ts:3]: translator exists on stardust

**Comment by haxxmaxx** [comment]
change the tests as well
